src/wallet/wallet.cpp

```diff
@@ -110,8 +110,8 @@ static void ReleaseWallet(CWallet* wallet)
     const std::string name = wallet->GetName();
     wallet->WalletLogPrintf("Releasing wallet\n");
     wallet->BlockUntilSyncedToCurrentChain();
-    wallet->Flush();
     wallet->m_chain_notifications_handler.reset();
+    WITH_LOCK(wallet->cs_wallet, wallet->Flush());
```

Comment:
Acquiring the cs_wallet lock here is unexpected. If this is really neccessary, I think the lock should be acquired closer to where it's used inside the Flush implementation, probably with a comment if the purpose of the lock isn't more obvious there

Reply:
> I think there is a misunderstanding here (maybe my own), but my understanding is that the point of BlockUntilSyncedToCurrentChain is just to ensure that if you make two RPC calls in a row the second RPC call is always aware of the result of first call. Ensuring that notification in the queue from the time of the call are processed is sufficient to do this, without trying to guarantee that the wallet tip and node tips are equal.

Yes, but mostly after BlockUntilSyncedToCurrentChain lock is acquired and there is not guarantee that notification is faster than this held, i.e. you can leave BlockUntilSyncedToCurrentChain and don't do what you want, you gain mutex before notification thread and you're not sync anything.

Reply:
Here after BlockUntilSyncedToCurrentChain is ensuring it's on current tip, the lock tries to guarantee that notification thread is finished before we flush and release memory.